### PR TITLE
libkmod: Simplify lookup_builtin_file()

### PR DESCRIPTION
**Pull request by lucasdemarchi**
Use cleanup attribute.

### CODE REVIEW

**Comment by evelikov** [comment]
Tl:Dr: `_cleanup_` coverage is at ~1%, with 1 instance added in ~450 patches since 2024/01/01 yet it contributes to 60% of false reports.

Longer version:
I guess we can step back and look at the bigger picture: how much do we use `_cleanup_` and does it outweigh the benefits of adding analyser like `scan-build` (aka clang analyze)?

 - looking around - we had 2 new instances over the last 1.5 years (both of those my me :stuck_out_tongue:) and third one back in 2018. In terms of count there are only 19 instances in the whole codebase, while they could easily be 10-20x as much.

- glance at recent~ish `scan-build` report shows 38 issues, 23 of which from our use of `_cleanup_` - the rest, at least in part, are genuine - writing to a `malloc(0)` block, passing `NULL to `_nonnull_` annotated API, etc

Objectively looking our `_cleanup_` use is at 1%, while it contributes ~60% of noise to a `scan-build` run. This doesn't seems worth it.

On the subjective side - it's sad and annoying to remove a _really_ cool feature. Personally, if `scan-build` properly supported the attribute, I would have used a lot more than in 2 out of 200+ patches. Alas that's not the case ...

Plus there is nothing blocking us from re-introducing it later on when scan-build gets wiser.

**Comment by lucasdemarchi** [comment]
I'm still of the opinion of scan-build needs to get wiser or we need to convince it of a proper way to understand it. Because it seems they are already even checking for improper use of the cleanup attribute. How the heck they don't understand what the attribute does?

You are creating a chicken and egg situation: it doesn't get used more because of X and X doesn't improve because it's not used enough. When something is that dumb, I prefer to simply disable it and use something smarter to cover the same thing. We can simply disable the unix.Malloc check and rely on valgrind for that.

Another alternative: try to convince the analyzer to stop bothering us with those false positives

**Comment by evelikov** [comment]
Fully agree that clang-analyze should get wiser - yet again I'm just the messenger here. The chicken and egg seems valid if clang devs see kmod (plus other users like systemd) as moderate-high interest/priority project. Hope I don't sound like a cold-hearted a** here - trying to be realistic.

Thanks for the `::suppress` trick - didn't knew it existed :bow: 

At a glance it seems that the allocation needs to be annotated, not the variable declaration. While moving the assignment/allocation on declaration works, we cannot reliably use it where we don't know the size early "enough" since we explicitly ban declaration after statements.

If we remove `-Wdeclaration-after-statement` (I'm in favour btw) this would help us move things as needed.

**Comment by lucasdemarchi** [comment]
See the last commit with the fixup where I return it to where it was. The problem is, like I said, even with those it seems like the leak check in clang-analyzer is not smart enough and has other problems. To the point I think we'd be better off turning off that specific check.

**Comment by evelikov** [comment]
Apologies - had the reply written on the day before and didn't refresh/check before posting :bow: 

Looking at the 23 mem leak reports we have:
 - attribute cleanup related 10
 - kmod_list ownership transfer 7
 - legitimate 6 - some of which in the tests

Suspect we could "fix" the kmod_list ones although it'll be alot of churn. Plus there is a 10+ year [old issue](https://github.com/llvm/llvm-project/issues/4260) about the cleanup attribute with limited activity.

So despite the legitimate catches, the noise is too much to have `unix.Malloc` enabled by default :sob:

**Review by evelikov — APPROVED** [review]
Thumbs up for reintroducing the `_cleanup_free_` instance. The clang/suppress patch can be deferred for a day when the false-positives are fewer.

**Comment by lucasdemarchi** [comment]
Yeah, I think I will just check for running with clang_analyzer and remove the attribute check.

**Comment by lucasdemarchi** [comment]
Applied, thanks
